tool/src/java/org/sakaiproject/gradebookng/tool/panels/AddGradeItemPanel.java

```diff
@@ -14,14 +27,82 @@
+
+    AjaxButton submit = new AjaxButton("submit") {
+      private static final long serialVersionUID = 1L;
+      @Override
+      public void onSubmit(AjaxRequestTarget target, Form<?> form) {
+        GbAssignment model =  (GbAssignment) form.getModelObject();
+        Assignment assignment = model.convert2Assignment();
+
+        boolean success = true;
+        try {
+          ((BasePage) getPage()).businessService.addAssignment(assignment);
+        } catch (AssignmentHasIllegalPointsException e) {
+          error(new ResourceModel("error.addgradeitem.points").getObject());
+          success = false;
+        } catch (ConflictingAssignmentNameException e) {
+          error(new ResourceModel("error.addgradeitem.title").getObject());
+          success = false;
+        } catch (ConflictingExternalIdException e) {
+          error(new ResourceModel("error.addgradeitem.exception").getObject());
+          success = false;
+        } catch (Exception e) {
+          error(new ResourceModel("error.addgradeitem.exception").getObject());
+          success = false;
+        }
+        if (success) {
+          getSession().info(new ResourceModel("notification.addgradeitem.success").getObject());
+          setResponsePage(getPage().getPageClass());
+        } else {
+          target.addChildren(form, FeedbackPanel.class);
+        }
+        
+      }
+    };
+		form.add(submit);
+
+    form.add(new AddGradeItemPanelContent("subComponents", model, categories));
+
+    FeedbackPanel feedback = new FeedbackPanel("addGradeFeedback") {
+      private static final long serialVersionUID = 1L;
 
-        Form<?> form = new Form("form");
+      @Override
+      protected Component newMessageDisplayComponent(final String id, final FeedbackMessage message) {
+        final Component newMessageDisplayComponent = super.newMessageDisplayComponent(id, message);
 
-		add(new AddGradeItemPanelContent("subComponents", new GbAssignmentModel(new GbAssignment())));
+        if(message.getLevel() == FeedbackMessage.ERROR ||
+                message.getLevel() == FeedbackMessage.DEBUG ||
+                message.getLevel() == FeedbackMessage.FATAL ||
+                message.getLevel() == FeedbackMessage.WARNING){
+          add(AttributeModifier.replace("class", "messageError"));
+          add(AttributeModifier.append("class", "feedback"));
+        } else if(message.getLevel() == FeedbackMessage.INFO){
+          add(AttributeModifier.replace("class", "messageSuccess"));
+          add(AttributeModifier.append("class", "feedback"));
+        }
 
+        return newMessageDisplayComponent;
+      }
+    };
+    feedback.setOutputMarkupId(true);
+    form.add(feedback);
 
+    AjaxButton cancel = new AjaxButton("cancel") {
+      private static final long serialVersionUID = 1L;
+      @Override
+      public void onSubmit(AjaxRequestTarget target, Form<?> form) {
+        window.close(target);
+      }
+    };
+    form.add(cancel);
+    add(form);
```

Comment:
Ah crap yeah.  Any preference what tabs/formatting we go with? I've found I have to change the IDE settings file to file from soft to hard tabs, 2-4 spaces.